Patch series to improve podvm image generation process

bpradipt opened the pull request:
This series adds support for creating a separate container image for podvm binaries and uses the same for creating podvm image.

bpradipt commented:
The gh-action needs to be updated to create all the images. I plan to handle this in a separate PR if it's fine.

snir911 commented on podvm/Dockerfile.podvm:
```diff
@@ -33,22 +33,23 @@ RUN if [ -n "${CAA_SRC}" ]; then \
-# Building binaries separately allows for reuse in case of error
-# with image building step OTOH it may increase layers size
-#RUN cd cloud-api-adaptor/podvm && \
-#     make binaries
```
I'd prefer to leave it but not a deal but fine with removing as well

bpradipt replied:
Removed it to avoid any confusion. But I can add commented out code for anyone wanting to rebuilding the binaries when creating the image

snir911 reviewed (approved):
Thanks, tested rhel on aws, worked well

stevenhorsman reviewed (commented):
I've run this locally to create an s390x ibmcloud based image (full disclosure - I've not tried running a peer pod from it though), so it is general good. I have a question in the code and then
I wonder if we should update the doc in https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/README.md as part of this to specify the extra step:
```
docker build -t podvm_binaries_s390x --build-arg ARCH=s390x --build-arg BUILDER_IMG=podvm_builder:latest --build-arg CLOUD_PROVIDER=ibmcloud -f Dockerfile.podvm_binaries .
```
and updated the podvm build to pass in `BINARIES_IMG` as well?

bpradipt commented on podvm/Dockerfile.podvm:
```diff
@@ -5,20 +5,20 @@
 # Builds pod vm image inside container
 #
 ARG BUILDER_IMG="quay.io/confidential-containers/podvm-builder-ubuntu"
+ARG BINARIES_IMG="quay.io/confidential-containers/podvm-binaries-centos"
```
Thanks for the catch. The above line is incorrect.
The binaries and  builder should be using the same distro although the builds are static unless it's rhel. But still

bpradipt commented:
> I've run this locally to create an s390x ibmcloud based image (full disclosure - I've not tried running a peer pod from it though), so it is general good. I have a question in the code and then I wonder if we should update the doc in https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/README.md as part of this to specify the extra step:
> 
> ```
> docker build -t podvm_binaries_s390x --build-arg ARCH=s390x --build-arg BUILDER_IMG=podvm_builder:latest --build-arg CLOUD_PROVIDER=ibmcloud -f Dockerfile.podvm_binaries .
> ```
> 
> and updated the podvm build to pass in `BINARIES_IMG` as well?

Thanks for mentioning this. I updated the README.

stevenhorsman commented on podvm/README.md:
````diff
@@ -73,48 +74,82 @@ $ docker build -t podvm_builder \
 
-Below is shown how to build an Ubuntu image for libvirt:
+```
+The build process can take significant time.
````
Sorry to nitpick, but I think this build is sensitive to architecture too? If so it might be worth adding a sentence about the option to setting `--build-arg ARCH`?

stevenhorsman commented on podvm/README.md:
```diff
@@ -73,48 +74,82 @@ $ docker build -t podvm_builder \
 
-In order to build the podvm image you should be using the corresponding dockerfile of the Linux distro for which the builder image was built. For example, if the builder image was built with *Dockerfile.podvm_builder.DISTRO* then you should use the *Dockerfile.podvm.DISTRO* to build the podvm image.
+Assuming you have built the podvm_builder image for Ubuntu as explained in the previous step,
+running the following command will build the image with the podvm binaries.
```
Sorry to nitpick, but I think this build is sensitive to architecture too? If so it might be worth adding a sentence about the option to setting `--build-arg ARCH`?

whitebrandy commented on podvm/Dockerfile.podvm_binaries:
```diff
@@ -0,0 +1,44 @@
+RUN tar czvf /podvm-binaries.tar.gz -C /src/cloud-api-adaptor/podvm/files usr/
+RUN tar czvf /pause-bundle.tar.gz -C /src/cloud-api-adaptor/podvm/files pause_bundle/
+
+FROM ubuntu:20.04
```
I am wondering if the distro specific build stage is necessary ? Can we use the minimal `FROM scratch` instead since we are only copying ?

bpradipt replied:
I explicitly added it to make it easier for running commands using the same image if needed. Let's keep it for sometime and we can change it going forward.

whitebrandy replied:
Sounds good. The advantage of scratch is it's a no-op. So, you actually save some cycles by avoiding creating a layer.

stevenhorsman commented on podvm/README.md:
```diff
@@ -73,48 +74,82 @@ $ docker build -t podvm_builder \
 
-This step will take several minutes to complete, mostly because `packer` will use the QEMU builder in emulation mode when running within container. If you are using podman then you can speed up QEMU by enabling native virtualization, i.e., passing the `--device=/dev/kvm` argument to enable KVM inside the container.
+This step will take several minutes to complete, mostly because `packer` will
+use the QEMU builder in emulation mode when running within container. If you
+are using podman then you can speed up QEMU by enabling native virtualization,
+i.e., passing the `--device=/dev/kvm` argument to enable KVM inside the
+container.
```
Nit: consider using the special GFM for Tips here?
```suggestion
use the QEMU builder in emulation mode when running within container.
> **Tip:** If you are using podman then you can speed up QEMU by enabling native
> virtualization, by passing the `--device=/dev/kvm` argument to enable KVM inside
> the container.
```
(I hope I got the formatted correct, but have no preview in the PR comments)

stevenhorsman commented on podvm/README.md:
```diff
@@ -73,48 +74,82 @@ $ docker build -t podvm_builder \
+container.
 
-Also beware that the process consume a bunch of memory and disk from the host. As a tip, if the build fail at the point QEMU was launched but packer couldn't connect via ssh then it might indicate lack of memory (try to increase the amount of memory if running on VM).
+Also beware that the process consume a bunch of memory and disk from the host.
+As a tip, if the build fail at the point QEMU was launched but packer couldn't
+connect via ssh then it might indicate lack of memory (try to increase the
+amount of memory if running on VM).
```
Nit: formatting and adding example error
```suggestion
> **Note:** Beware that the process consume a bunch of memory and disk from the host.
If the build fails at the point QEMU was launched but packer couldn't
connect via ssh, with an error similar to:
> ```
> Build 'qemu.ubuntu' errored after 5 minutes 57 seconds: Timeout waiting for SSH.
> ```
> then it might indicate lack of memory, so try to increase the amount of memory if running on VM.
```

stevenhorsman reviewed (approved):
Thanks very much for adding the updated documentation, it's generally really good, so I feel quite guilty, but I've made a few suggestions. Nothing major though.

bpradipt commented:
> Thanks very much for adding the updated documentation, it's generally really good, so I feel quite guilty, but I've made a few suggestions. Nothing major though.

Thanks for your suggestions. I have updated the README

stevenhorsman reviewed (approved):
Looks good to me. Thanks for the updates!
